apps/jupyter/extensions/notebook-runner/src/packages.ts

```diff
@@ -0,0 +1,208 @@
+    await importBasePackages(kernel);
+    await installOtter(kernel);
+    console.debug("Finished installing packages", kernel.id);
+
+    if (
+      notebookPath === EmbeddedPythonCallbacks.studentTaskLocation ||
+      notebookPath === EmbeddedPythonCallbacks.studentTaskLocation.slice(1)
+    ) {
+      console.debug(
+        "Opened student notebook - copying notebook to the virtual filesystem to make tests available..",
+      );
+
+      let notebook: Contents.IModel | null = null;
+      try {
+        notebook = await contentsManager.get(notebookPath, { content: true });
+      } catch (error) {
+        throw new Error("Error reading ran notebook:" + error);
+      }
+
+      await writeJsonToVirtualFilesystem(
+        kernel,
+        EmbeddedPythonCallbacks.studentTaskLocation,
+        notebook.content,
+      );
+      console.debug("Finished copying notebook to the virtual filesystem");
+
+      await kernel.requestExecute({
+        code: `os.chdir("/student")`,
+      }).done;
+    }
+  };
+
+const trackSession = (
+  contentsManager: ContentsManager,
+  sessionContext: ISessionContext,
+  notebookPath: string,
+): Promise<void> =>
+  addKernelListeners(
+    sessionContext,
+    autoInstallPackages(contentsManager, notebookPath),
+  );
+
+export const preInstallPackages = async (
+  _app: JupyterFrontEnd,
+  contentsManager: ContentsManager,
+  notebookTracker: INotebookTracker,
+): Promise<void> => {
+  notebookTracker.forEach((panel) =>
+    trackSession(contentsManager, panel.sessionContext, panel.context.path),
+  );
+
+  notebookTracker.widgetAdded.connect((_, panel) => {
+    // get path of notebook
+    const notebookPath = panel.context.path;
+
+    trackSession(contentsManager, panel.sessionContext, notebookPath);
+  });
+};
+
+export const importBasePackages = async (
+  kernel: IKernelConnection,
+): Promise<void> => {
+  console.log("Importing builtin libraries..");
+  await kernel.requestExecute(
+    {
+      code: `
+        import micropip, sys, types, os, tempfile, asyncio, shutil, base64
+        from urllib.parse import unquote
+        from pathlib import Path
+      `,
+    },
+    true,
+  ).done;
+};
+
+export const installOtter = async (
+  kernel: IKernelConnection,
+): Promise<void> => {
+  console.log("Installing Otter Grader..");
+  await kernel.requestExecute(
+    {
+      code: `
+await micropip.install("/jupyter/pypi/otter_grader-6.1.3-py3-none-any.whl")
+      `,
+    },
+    true,
+  ).done;
+};
+
+export const installNbConvert = async (
+  kernel: IKernelConnection,
+): Promise<void> => {
+  console.debug("Installing nbconvert..");
+  await kernel.requestExecute(
+    {
+      code: `
+        # Micropip by default installs all dependencies including tornado which is not mandatory,
+        # especially for the codepaths we need.
+        # The list of dependencies can be seen here: https://github.com/jupyter/nbconvert/blob/main/pyproject.toml#L28
```

Comment:
This is in response to `deps=False`, but the connection between the two is never made explicit. Can we make that connection explicit, so that somebody doesn't have to scratch their head like I just did, wondering "why are you telling me this?" 😆